fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1278,7 +1284,8 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
             # Free full params. As a special case, we don't free the full params
             # when in a ``no_sync`` context (as inversely indicated by
             # ``self._require_backward_grad_sync``), since the params will not
-            # get updated before the next forward.
+            # get updated before the next forward. This saves networking
+            # bandwidth but uses more GPU memory.
             self._free_full_params([param])
```

Comment:
I'm wondering whether full params could be freed and local grad shard is kept in 'no_sync' mode, in this case, 'no_sync' can save GPU memory, right?

Reply:
Good question. In no sync mode, it is both full params and full grads that are consuming the GPU memory.

- We keep full params since the params are not updated while we accumulate the grads
- We keep the full grads since we are not reducing the scatter them (i.e. shard them)

In this new change, we free and reshard full params and shard grads and accumulate into grad shards. So both of the memory overhead above are avoided at the cost of networking overhead.